File: src/HotRepl.Core/Protocol/Messages.cs
using System;
using HotRepl;
using HotRepl.Evaluator;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace HotRepl.Protocol;

/// <summary>Wire-format type discriminants for all protocol messages.</summary>
internal static class MessageType
{
    // Inbound
    public const string Eval = "eval";
    public const string Cancel = "cancel";
    public const string Reset = "reset";
    public const string Ping = "ping";
    public const string Complete = "complete";
    public const string Subscribe = "subscribe";
    public const string SelectEvaluator = "select_evaluator";
    public const string ControlAuth = "control_auth";
    public const string LeaseAcquire = "lease_acquire";
    public const string CommandDescribe = "command_describe";
    public const string CommandCall = "command_call";
    public const string JobStatus = "job_status";
    public const string JobCancel = "job_cancel";

    // Outbound
    public const string Handshake = "handshake";
    public const string EvalResult = "eval_result";
    public const string EvalError = "eval_error";
    public const string ResetResult = "reset_result";
    public const string Pong = "pong";
    public const string CompleteResult = "complete_result";
    public const string SubscribeResult = "subscribe_result";
    public const string SubscribeError = "subscribe_error";
    public const string AssemblyReload = "assembly_reload";
    public const string SelectEvaluatorResult = "select_evaluator_result";
    public const string SelectEvaluatorError = "select_evaluator_error";
    public const string ControlAuthResult = "control_auth_result";
    public const string LeaseAcquireResult = "lease_acquire_result";
    public const string CommandDescribeResult = "command_describe_result";
    public const string CommandResult = "command_result";
    public const string CommandError = "command_error";
    public const string CommandAccepted = "command_accepted";
    public const string JobStatusResult = "job_status_result";
    public const string JobResult = "job_result";
    public const string JobCancelResult = "job_cancel_result";
    public const string JobEvent = "job_event";
}

/// <summary>Eval error kind discriminants.</summary>
internal static class ErrorKind
{
    public const string Compile = "compile";
    public const string Runtime = "runtime";
    public const string Timeout = "timeout";
    public const string Cancelled = "cancelled";
    public const string Unsupported = "unsupported";
}

// ── Inbound ───────────────────────────────────────────────────────────────────

internal sealed class EvalMessage
{
    [JsonProperty("type")] public string Type { get; init; } = MessageType.Eval;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("code")] public string Code { get; set; } = string.Empty;
    [JsonProperty("timeoutMs")] public int TimeoutMs { get; set; }
}

internal sealed class CancelMessage
{
    [JsonProperty("type")] public string Type { get; init; } = MessageType.Cancel;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
}

internal sealed class ResetMessage
{
    [JsonProperty("type")] public string Type { get; init; } = MessageType.Reset;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
}

internal sealed class PingMessage
{
    [JsonProperty("type")] public string Type { get; init; } = MessageType.Ping;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
}

internal sealed class CompleteMessage
{
    [JsonProperty("type")] public string Type { get; init; } = MessageType.Complete;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("code")] public string Code { get; set; } = string.Empty;
    [JsonProperty("cursorPos")] public int CursorPos { get; set; } = -1;
}

internal sealed class SubscribeMessage
{
    [JsonProperty("type")] public string Type { get; init; } = MessageType.Subscribe;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("code")] public string Code { get; set; } = string.Empty;
    [JsonProperty("intervalFrames")] public int IntervalFrames { get; set; } = 1;
    [JsonProperty("onChange")] public bool OnChange { get; set; }
    [JsonProperty("limit")] public int Limit { get; set; }
    [JsonProperty("timeoutMs")] public int TimeoutMs { get; set; }
}

internal sealed class SelectEvaluatorMessage
{
    [JsonProperty("type")] public string Type { get; init; } = MessageType.SelectEvaluator;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("evaluator")] public string Evaluator { get; set; } = string.Empty;
}

internal sealed class ControlAuthMessage
{
    [JsonProperty("type")] public string Type { get; init; } = MessageType.ControlAuth;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("token")] public string? Token { get; set; }
}

internal sealed class LeaseAcquireMessage
{
    [JsonProperty("type")] public string Type { get; init; } = MessageType.LeaseAcquire;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("sessionId")] public string SessionId { get; set; } = string.Empty;
    [JsonProperty("clientName")] public string ClientName { get; set; } = string.Empty;
}

internal sealed class CommandDescribeMessage
{
    [JsonProperty("type")] public string Type { get; init; } = MessageType.CommandDescribe;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
}

internal sealed class CommandCallMessage
{
    [JsonProperty("type")] public string Type { get; init; } = MessageType.CommandCall;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("leaseId")] public string? LeaseId { get; set; }
    [JsonProperty("name")] public string Name { get; set; } = string.Empty;
    [JsonProperty("args")] public JObject Args { get; set; } = new();
    [JsonProperty("timeoutMs")] public int TimeoutMs { get; set; }
    [JsonProperty("idempotencyKey")] public string? IdempotencyKey { get; set; }
}

internal sealed class JobStatusMessage
{
    [JsonProperty("type")] public string Type { get; init; } = MessageType.JobStatus;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("leaseId")] public string? LeaseId { get; set; }
    [JsonProperty("jobId")] public string JobId { get; set; } = string.Empty;
}

internal sealed class JobCancelMessage
{
    [JsonProperty("type")] public string Type { get; init; } = MessageType.JobCancel;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("leaseId")] public string? LeaseId { get; set; }
    [JsonProperty("jobId")] public string JobId { get; set; } = string.Empty;
}

// ── Outbound ──────────────────────────────────────────────────────────────────

internal sealed class HandshakeMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.Handshake;
    [JsonProperty("version")] public string Version { get; set; } = string.Empty;
    [JsonProperty("csharpVersion")] public string CsharpVersion { get; set; } = "7.x";
    [JsonProperty("defaultUsings")] public string[] DefaultUsings { get; set; } = Array.Empty<string>();
    [JsonProperty("helpers")] public string[] Helpers { get; set; } = Array.Empty<string>();
    [JsonProperty("evaluator")] public EvaluatorCapabilities? Evaluator { get; set; }
    [JsonProperty("host")] public HostInfo? Host { get; set; }
    [JsonProperty("availableEvaluators")] public string[] AvailableEvaluators { get; set; } = Array.Empty<string>();
}

internal sealed class EvalResultMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.EvalResult;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("hasValue")] public bool HasValue { get; set; }
    [JsonProperty("value")] public string? Value { get; set; }
    [JsonProperty("valueType")] public string? ValueType { get; set; }
    [JsonProperty("stdout")] public string? Stdout { get; set; }
    [JsonProperty("durationMs")] public long DurationMs { get; set; }
}

internal sealed class EvalErrorMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.EvalError;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("errorKind")] public string ErrorKind { get; set; } = string.Empty;
    [JsonProperty("message")] public string Message { get; set; } = string.Empty;
    [JsonProperty("stackTrace")] public string? StackTrace { get; set; }
}

internal sealed class ResetResultMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.ResetResult;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("success")] public bool Success { get; set; }
}

internal sealed class SelectEvaluatorResultMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.SelectEvaluatorResult;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("success")] public bool Success { get; set; }
    [JsonProperty("evaluator")] public string Evaluator { get; set; } = string.Empty;
}

internal sealed class SelectEvaluatorErrorMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.SelectEvaluatorError;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("errorKind")] public string ErrorKind { get; set; } = string.Empty;
    [JsonProperty("message")] public string Message { get; set; } = string.Empty;
}

internal sealed class PongMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.Pong;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
}

internal sealed class CompleteResultMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.CompleteResult;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("completions")] public string[] Completions { get; set; } = Array.Empty<string>();
    [JsonProperty("durationMs")] public long DurationMs { get; set; }
}

internal sealed class SubscribeResultMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.SubscribeResult;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("seq")] public int Seq { get; set; }
    [JsonProperty("hasValue")] public bool HasValue { get; set; }
    [JsonProperty("value")] public string? Value { get; set; }
    [JsonProperty("valueType")] public string? ValueType { get; set; }
    [JsonProperty("durationMs")] public long DurationMs { get; set; }
    [JsonProperty("final")] public bool Final { get; set; }
}

internal sealed class SubscribeErrorMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.SubscribeError;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("seq")] public int Seq { get; set; }
    [JsonProperty("errorKind")] public string ErrorKind { get; set; } = string.Empty;
    [JsonProperty("message")] public string Message { get; set; } = string.Empty;
    [JsonProperty("final")] public bool Final { get; set; }
}

internal sealed class AssemblyReloadMessage
{
    [JsonProperty("type")] public string Type => MessageType.AssemblyReload;
    [JsonProperty("assembly")] public string? Assembly { get; set; }
    [JsonProperty("message")] public string? Message { get; set; }
}

internal sealed class ControlAuthResultMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.ControlAuthResult;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("ok")] public bool Ok { get; set; }
    [JsonProperty("sessionId")] public string? SessionId { get; set; }
    [JsonProperty("error")] public ControlErrorMessage? Error { get; set; }
}

internal sealed class LeaseAcquireResultMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.LeaseAcquireResult;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("ok")] public bool Ok { get; set; }
    [JsonProperty("leaseId")] public string? LeaseId { get; set; }
    [JsonProperty("error")] public ControlErrorMessage? Error { get; set; }
}

internal sealed class CommandDescriptorMessage
{
    [JsonProperty("name")] public string Name { get; set; } = string.Empty;
    [JsonProperty("version")] public int Version { get; set; }
    [JsonProperty("kind")] public string Kind { get; set; } = string.Empty;
    [JsonProperty("mutatesState")] public bool MutatesState { get; set; }
    [JsonProperty("argsSchema")] public JObject ArgsSchema { get; set; } = new();
    [JsonProperty("resultSchema")] public JObject ResultSchema { get; set; } = new();
}

internal sealed class CommandDescribeResultMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.CommandDescribeResult;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("commands")] public CommandDescriptorMessage[] Commands { get; set; } = Array.Empty<CommandDescriptorMessage>();
}

internal sealed class CommandResultMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.CommandResult;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("status")] public string Status { get; set; } = string.Empty;
    [JsonProperty("result")] public JObject Result { get; set; } = new();
    [JsonProperty("artifacts")] public ArtifactRefMessage[] Artifacts { get; set; } = Array.Empty<ArtifactRefMessage>();
    [JsonProperty("diagnostics")] public ControlErrorMessage[] Diagnostics { get; set; } = Array.Empty<ControlErrorMessage>();
}

internal sealed class CommandErrorMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.CommandError;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("status")] public string Status { get; set; } = string.Empty;
    [JsonProperty("error")] public ControlErrorMessage Error { get; set; } = new();
    [JsonProperty("diagnostics")] public ControlErrorMessage[] Diagnostics { get; set; } = Array.Empty<ControlErrorMessage>();
}

internal sealed class CommandAcceptedMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.CommandAccepted;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("jobId")] public string JobId { get; set; } = string.Empty;
    [JsonProperty("state")] public string State { get; set; } = string.Empty;
}

internal sealed class JobStatusResultMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.JobStatusResult;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("jobId")] public string JobId { get; set; } = string.Empty;
    [JsonProperty("state")] public string State { get; set; } = string.Empty;
    [JsonProperty("progress")] public JObject? Progress { get; set; }
}

internal sealed class JobResultMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.JobResult;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("jobId")] public string JobId { get; set; } = string.Empty;
    [JsonProperty("state")] public string State { get; set; } = string.Empty;
    [JsonProperty("status")] public string Status { get; set; } = string.Empty;
    [JsonProperty("result")] public JObject Result { get; set; } = new();
    [JsonProperty("artifacts")] public ArtifactRefMessage[] Artifacts { get; set; } = Array.Empty<ArtifactRefMessage>();
    [JsonProperty("diagnostics")] public ControlErrorMessage[] Diagnostics { get; set; } = Array.Empty<ControlErrorMessage>();
}

internal sealed class JobCancelResultMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.JobCancelResult;
    [JsonProperty("id")] public string Id { get; set; } = string.Empty;
    [JsonProperty("accepted")] public bool Accepted { get; set; }
    [JsonProperty("state")] public string State { get; set; } = string.Empty;
}

internal sealed class JobEventMessage
{
    [JsonProperty("type")] public string Type { get; } = MessageType.JobEvent;
    [JsonProperty("jobId")] public string JobId { get; set; } = string.Empty;
    [JsonProperty("sequence")] public long Sequence { get; set; }
    [JsonProperty("state")] public string State { get; set; } = string.Empty;
    [JsonProperty("progress")] public JObject? Progress { get; set; }
    [JsonProperty("message")] public string? Message { get; set; }
}

internal sealed class ArtifactRefMessage
{
    [JsonProperty("logicalName")] public string LogicalName { get; set; } = string.Empty;
    [JsonProperty("uri")] public string Uri { get; set; } = string.Empty;
    [JsonProperty("path")] public string? Path { get; set; }
    [JsonProperty("contentType")] public string ContentType { get; set; } = string.Empty;
    [JsonProperty("byteSize")] public long ByteSize { get; set; }
    [JsonProperty("sha256")] public string Sha256 { get; set; } = string.Empty;
    [JsonProperty("finalized")] public bool Finalized { get; set; }
}

internal sealed class ControlErrorMessage
{
    [JsonProperty("kind")] public string Kind { get; set; } = string.Empty;
    [JsonProperty("code")] public string Code { get; set; } = string.Empty;
    [JsonProperty("message")] public string Message { get; set; } = string.Empty;
    [JsonProperty("retryable")] public bool Retryable { get; set; }
    [JsonProperty("details")] public JObject? Details { get; set; }
}

File: packages/protocol/src/messages.ts
import { type Static, Type } from "typebox";
import type { TLiteral } from "typebox";
import type { ErrorKind } from "./error-kinds";
import { ERROR_KINDS } from "./error-kinds";
import { type HandshakeMessage, HandshakeMessageSchema } from "./handshake";
import { MESSAGE_TYPES } from "./message-types";
// Preserve literal tuple types for Type.Union — plain .map() loses tuple info in TypeBox 1.x
type LiteralTuple<T extends readonly string[]> = T extends readonly [
  infer Head extends string,
  ...infer Tail extends readonly string[],
] ? [TLiteral<Head>, ...LiteralTuple<Tail>]
  : [];
type ErrorKindLiteralTuple = LiteralTuple<typeof ERROR_KINDS>;
const ERROR_KIND_LITERALS = ERROR_KINDS.map((k) =>
  Type.Literal(k)
) as unknown as ErrorKindLiteralTuple;

export type JsonObject = Record<string, unknown>;

// ── Shared types ──────────────────────────────────────────────────────────────

/** Opaque JSON object — used for command args and inline JSON schemas */
export const JsonObjectSchema = Type.Record(Type.String(), Type.Unknown());

/** Unified error envelope used in every failure response */
export const ErrorEnvelopeSchema = Type.Object(
  {
    kind: Type.Union(ERROR_KIND_LITERALS),
    code: Type.String(),
    message: Type.String(),
    retryable: Type.Boolean(),
    details: Type.Optional(Type.Unknown()),
  },
  { additionalProperties: false },
);
export type HotReplErrorEnvelope = Static<typeof ErrorEnvelopeSchema>;

/** Named reference to a file artifact produced by a command */
export const ArtifactRefSchema = Type.Object(
  {
    uri: Type.String(),
    path: Type.Optional(Type.String()),
    sha256: Type.String(),
    byteSize: Type.Number(),
    contentType: Type.String(),
    finalized: Type.Boolean(),
  },
  { additionalProperties: false },
);
export type ArtifactRef = Static<typeof ArtifactRefSchema>;

/** Summary of a registered typed command */
export const CommandSummarySchema = Type.Object(
  {
    name: Type.String(),
    majorVersion: Type.Number(),
    kind: Type.Union([Type.Literal("sync"), Type.Literal("job")]),
    mutatesState: Type.Boolean(),
  },
  { additionalProperties: false },
);
export type CommandSummary = Static<typeof CommandSummarySchema>;

/** Full descriptor for a registered typed command, including I/O schemas.
 *  Fields duplicated from CommandSummary deliberately (avoids anyOf/allOf in
 *  the exported JSON Schema, which makes the docs harder to read). */
export const CommandDescriptorSchema = Type.Object(
  {
    name: Type.String(),
    majorVersion: Type.Number(),
    kind: Type.Union([Type.Literal("sync"), Type.Literal("job")]),
    mutatesState: Type.Boolean(),
    inputSchema: JsonObjectSchema,
    outputSchema: JsonObjectSchema,
    artifactsSchema: JsonObjectSchema,
    cancellation: Type.Optional(Type.String()),
  },
  { additionalProperties: false },
);
export type CommandDescriptor = Static<typeof CommandDescriptorSchema>;

/** One eval or command entry in the journal */
export const JournalEntrySchema = Type.Object(
  {
    id: Type.String(),
    kind: Type.Union([Type.Literal("eval"), Type.Literal("command")]),
    name: Type.Optional(Type.String()),
    code: Type.Optional(Type.String()),
    success: Type.Boolean(),
    durationMs: Type.Number(),
    errorKind: Type.Optional(Type.Union(ERROR_KIND_LITERALS)),
    timestamp: Type.String(),
  },
  { additionalProperties: false },
);
export type JournalEntry = Static<typeof JournalEntrySchema>;

// ── Server-sent messages ──────────────────────────────────────────────────────

export const EvalResultMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.evalResult),
    id: Type.String(),
    hasValue: Type.Boolean(),
    value: Type.Optional(Type.Unknown()),
    valueType: Type.Optional(Type.String()),
    truncated: Type.Optional(Type.Boolean()),
    truncatedBytes: Type.Optional(Type.Number()),
    stdout: Type.Optional(Type.String()),
    durationMs: Type.Number(),
  },
  { additionalProperties: false },
);
export type EvalResultMessage = Static<typeof EvalResultMessageSchema>;

export const EvalErrorMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.evalError),
    id: Type.String(),
    error: ErrorEnvelopeSchema,
  },
  { additionalProperties: false },
);
export type EvalErrorMessage = Static<typeof EvalErrorMessageSchema>;

export const CompleteResultMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.completeResult),
    id: Type.String(),
    completions: Type.Array(Type.String()),
    durationMs: Type.Number(),
  },
  { additionalProperties: false },
);
export type CompleteResultMessage = Static<typeof CompleteResultMessageSchema>;

export const ResetResultMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.resetResult),
    id: Type.String(),
    success: Type.Boolean(),
  },
  { additionalProperties: false },
);
export type ResetResultMessage = Static<typeof ResetResultMessageSchema>;

export const SubscribeResultMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.subscribeResult),
    id: Type.String(),
    seq: Type.Number(),
    hasValue: Type.Boolean(),
    value: Type.Optional(Type.Unknown()),
    valueType: Type.Optional(Type.String()),
    truncated: Type.Optional(Type.Boolean()),
    truncatedBytes: Type.Optional(Type.Number()),
    durationMs: Type.Number(),
    final: Type.Boolean(),
  },
  { additionalProperties: false },
);
export type SubscribeResultMessage = Static<typeof SubscribeResultMessageSchema>;

export const SubscribeErrorMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.subscribeError),
    id: Type.String(),
    seq: Type.Number(),
    error: ErrorEnvelopeSchema,
    final: Type.Boolean(),
  },
  { additionalProperties: false },
);
export type SubscribeErrorMessage = Static<typeof SubscribeErrorMessageSchema>;

export const SessionEvictedMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.sessionEvicted),
    reason: Type.String(),
    by: Type.Optional(
      Type.Object({ clientName: Type.Optional(Type.String()) }, { additionalProperties: false }),
    ),
  },
  { additionalProperties: false },
);
export type SessionEvictedMessage = Static<typeof SessionEvictedMessageSchema>;

export const ProtocolErrorMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.error),
    id: Type.Optional(Type.String()),
    error: ErrorEnvelopeSchema,
  },
  { additionalProperties: false },
);
export type ProtocolErrorMessage = Static<typeof ProtocolErrorMessageSchema>;

export const CommandsListResultMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.commandsListResult),
    id: Type.String(),
    commands: Type.Array(CommandSummarySchema),
    since: Type.Optional(Type.String()),
  },
  { additionalProperties: false },
);
export type CommandsListResultMessage = Static<typeof CommandsListResultMessageSchema>;

export const CommandDescribeResultMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.commandDescribeResult),
    id: Type.String(),
    descriptor: CommandDescriptorSchema,
  },
  { additionalProperties: false },
);
export type CommandDescribeResultMessage = Static<typeof CommandDescribeResultMessageSchema>;

export const CommandResultMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.commandResult),
    id: Type.String(),
    status: Type.Union([Type.Literal("ok"), Type.Literal("failed")]),
    output: Type.Optional(Type.Unknown()),
    artifacts: Type.Record(Type.String(), ArtifactRefSchema),
    error: Type.Optional(ErrorEnvelopeSchema),
    durationMs: Type.Number(),
  },
  { additionalProperties: false },
);
export type CommandResultMessage = Static<typeof CommandResultMessageSchema>;

export const JobAcceptedMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.jobAccepted),
    id: Type.String(),
    jobId: Type.String(),
    state: Type.Literal("running"),
  },
  { additionalProperties: false },
);
export type JobAcceptedMessage = Static<typeof JobAcceptedMessageSchema>;

export const JobStatusResultMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.jobStatusResult),
    id: Type.String(),
    jobId: Type.String(),
    state: Type.Literal("running"),
    progress: Type.Optional(Type.Unknown()),
    error: Type.Optional(ErrorEnvelopeSchema),
  },
  { additionalProperties: false },
);
export type JobStatusResultMessage = Static<typeof JobStatusResultMessageSchema>;

export const JobResultMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.jobResult),
    id: Type.String(),
    jobId: Type.String(),
    state: Type.Union([
      Type.Literal("done"),
      Type.Literal("failed"),
      Type.Literal("cancelled"),
    ]),
    status: Type.Union([Type.Literal("ok"), Type.Literal("failed")]),
    output: Type.Optional(Type.Unknown()),
    artifacts: Type.Record(Type.String(), ArtifactRefSchema),
    error: Type.Optional(ErrorEnvelopeSchema),
    durationMs: Type.Number(),
  },
  { additionalProperties: false },
);
export type JobResultMessage = Static<typeof JobResultMessageSchema>;

export const JobCancelResultMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.jobCancelResult),
    id: Type.String(),
    jobId: Type.String(),
    accepted: Type.Boolean(),
    state: Type.Union([
      Type.Literal("running"),
      Type.Literal("done"),
      Type.Literal("failed"),
      Type.Literal("cancelled"),
    ]),
  },
  { additionalProperties: false },
);
export type JobCancelResultMessage = Static<typeof JobCancelResultMessageSchema>;

export const JournalQueryResultMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.journalQueryResult),
    id: Type.String(),
    entries: Type.Array(JournalEntrySchema),
  },
  { additionalProperties: false },
);
export type JournalQueryResultMessage = Static<typeof JournalQueryResultMessageSchema>;

/** Sent by the server when a game assembly is hot-reloaded. The SDK surfaces this via
 *  Session.onAssemblyReload and invalidates its cached command catalog/descriptors. */
export const AssemblyReloadMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.assemblyReload),
    assembly: Type.Optional(Type.String()),
    message: Type.String(),
  },
  { additionalProperties: false },
);
export type AssemblyReloadMessage = Static<typeof AssemblyReloadMessageSchema>;

// ── ServerMessage union ───────────────────────────────────────────────────────

export const ServerMessageSchema = Type.Union([
  HandshakeMessageSchema,
  EvalResultMessageSchema,
  EvalErrorMessageSchema,
  CompleteResultMessageSchema,
  ResetResultMessageSchema,
  SubscribeResultMessageSchema,
  SubscribeErrorMessageSchema,
  SessionEvictedMessageSchema,
  ProtocolErrorMessageSchema,
  CommandsListResultMessageSchema,
  CommandDescribeResultMessageSchema,
  CommandResultMessageSchema,
  JobAcceptedMessageSchema,
  JobStatusResultMessageSchema,
  JobResultMessageSchema,
  JobCancelResultMessageSchema,
  JournalQueryResultMessageSchema,
  AssemblyReloadMessageSchema,
]);

export type ServerMessage =
  | HandshakeMessage
  | EvalResultMessage
  | EvalErrorMessage
  | CompleteResultMessage
  | ResetResultMessage
  | SubscribeResultMessage
  | SubscribeErrorMessage
  | SessionEvictedMessage
  | ProtocolErrorMessage
  | CommandsListResultMessage
  | CommandDescribeResultMessage
  | CommandResultMessage
  | JobAcceptedMessage
  | JobStatusResultMessage
  | JobResultMessage
  | JobCancelResultMessage
  | JournalQueryResultMessage
  | AssemblyReloadMessage;

// ── Client-sent messages ──────────────────────────────────────────────────────

export const EvalMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.eval),
    id: Type.String(),
    code: Type.String(),
    timeoutMs: Type.Optional(Type.Number()),
  },
  { additionalProperties: false },
);
export type EvalMessage = Static<typeof EvalMessageSchema>;

export const CompleteMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.complete),
    id: Type.String(),
    code: Type.String(),
    cursor: Type.Optional(Type.Number()),
  },
  { additionalProperties: false },
);
export type CompleteMessage = Static<typeof CompleteMessageSchema>;

export const ResetMessageSchema = Type.Object(
  { type: Type.Literal(MESSAGE_TYPES.reset), id: Type.String() },
  { additionalProperties: false },
);
export type ResetMessage = Static<typeof ResetMessageSchema>;

export const SubscribeMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.subscribe),
    id: Type.String(),
    code: Type.String(),
    intervalFrames: Type.Optional(Type.Number()),
    onChange: Type.Optional(Type.Boolean()),
    limit: Type.Optional(Type.Number()),
    timeoutMs: Type.Optional(Type.Number()),
  },
  { additionalProperties: false },
);
export type SubscribeMessage = Static<typeof SubscribeMessageSchema>;

/** Cancels an active eval or subscription by its request id. Sent by the SDK via
 *  Session.cancel(targetId), and automatically when a watch() iterator stops early. */
export const CancelMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.cancel),
    id: Type.String(),
    targetId: Type.String(),
  },
  { additionalProperties: false },
);
export type CancelMessage = Static<typeof CancelMessageSchema>;

export const CommandsListMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.commandsList),
    id: Type.String(),
    since: Type.Optional(Type.String()),
  },
  { additionalProperties: false },
);
export type CommandsListMessage = Static<typeof CommandsListMessageSchema>;

export const CommandDescribeMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.commandDescribe),
    id: Type.String(),
    name: Type.String(),
  },
  { additionalProperties: false },
);
export type CommandDescribeMessage = Static<typeof CommandDescribeMessageSchema>;

/** args must be a JSON object — the C# runtime deserializes into JObject. */
export const CommandCallMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.commandCall),
    id: Type.String(),
    name: Type.String(),
    args: JsonObjectSchema,
    timeoutMs: Type.Optional(Type.Number()),
  },
  { additionalProperties: false },
);
export type CommandCallMessage = Static<typeof CommandCallMessageSchema>;

export const JobStatusMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.jobStatus),
    id: Type.String(),
    jobId: Type.String(),
  },
  { additionalProperties: false },
);
export type JobStatusMessage = Static<typeof JobStatusMessageSchema>;

export const JobCancelMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.jobCancel),
    id: Type.String(),
    jobId: Type.String(),
  },
  { additionalProperties: false },
);
export type JobCancelMessage = Static<typeof JobCancelMessageSchema>;

export const JournalQueryMessageSchema = Type.Object(
  {
    type: Type.Literal(MESSAGE_TYPES.journalQuery),
    id: Type.String(),
    kind: Type.Optional(Type.Union([Type.Literal("eval"), Type.Literal("command")])),
    limit: Type.Optional(Type.Number()),
  },
  { additionalProperties: false },
);
export type JournalQueryMessage = Static<typeof JournalQueryMessageSchema>;

export type ClientMessage =
  | EvalMessage
  | CompleteMessage
  | ResetMessage
  | SubscribeMessage
  | CancelMessage
  | CommandsListMessage
  | CommandDescribeMessage
  | CommandCallMessage
  | JobStatusMessage
  | JobCancelMessage
  | JournalQueryMessage;
